raw-spans-grouper/raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java

```diff
@@ -88,56 +95,90 @@ public void punctuate(long timestamp) {
     cancellable.cancel();
 
-    Long emitTs = traceEmitTriggerStore.get(key);
-    if (null == emitTs) {
+    TraceState traceState = traceStateStore.get(key);
+    if (null == traceState
+        || null == traceState.getSpanIds()
+        || traceState.getSpanIds().isEmpty()) {
       /*
        todo - debug why this happens .
        Typically seen when punctuators are created via {@link RawSpansGroupingTransformer.restorePunctuators}
       */
       logger.warn(
-          "emitTs for tenant_id=[{}], trace_id=[{}] is missing.",
+          "TraceState for tenant_id=[{}], trace_id=[{}] is missing.",
           key.getTenantId(),
           HexUtils.getHex(key.getTraceId()));
       return;
     }
 
+    long emitTs = traceState.getEmitTs();
     if (emitTs <= timestamp) {
       // we can emit this trace so just delete the entry for this 'key'
-      traceEmitTriggerStore.delete(key);
       // Implies that no new spans for the trace have arrived within the last
       // 'groupingWindowTimeoutMs' interval
       // so the trace can be finalized and emitted
-      ValueAndTimestamp<RawSpans> agg = inflightTraceStore.delete(key);
-      ByteBuffer traceId;
-      String tenantId;
-      if (agg != null && !agg.value().getRawSpans().isEmpty()) {
-        RawSpan firstSpan = agg.value().getRawSpans().get(0);
-        traceId = firstSpan.getTraceId();
-        tenantId = firstSpan.getCustomerId();
-
-        List<RawSpan> rawSpanList = agg.value().getRawSpans();
-        recordSpansPerTrace(rawSpanList.size(), List.of(Tag.of("tenant_id", tenantId)));
-        Timestamps timestamps = trackEndToEndLatencyTimestamps(timestamp, agg);
-        StructuredTrace trace =
-            StructuredTraceBuilder.buildStructuredTraceFromRawSpans(
-                rawSpanList, traceId, tenantId, timestamps);
+      traceStateStore.delete(key);
+
+      ByteBuffer traceId = traceState.getTraceId();
+      String tenantId = traceState.getTenantId();
+
+      TreeSet<ByteBuffer> spanIdsSet =
+          new TreeSet<>(
+              (spanId1, spanId2) -> {
+                // the sort order for this set is determined by comparing the corresponding {@code
+                // SpanIdentifier}
+                Serde<SpanIdentity> serde = (Serde<SpanIdentity>) context.keySerde();
+                byte[] spanIdentityBytes1 =
+                    serde.serializer().serialize("", new SpanIdentity(tenantId, traceId, spanId1));
+                byte[] spanIdentityBytes2 =
+                    serde.serializer().serialize("", new SpanIdentity(tenantId, traceId, spanId2));
+                return Bytes.BYTES_LEXICO_COMPARATOR.compare(
+                    spanIdentityBytes1, spanIdentityBytes2);
+              });
+      spanIdsSet.addAll(traceState.getSpanIds());
+
+      List<RawSpan> rawSpanList = new ArrayList<>();
+      try (KeyValueIterator<Windowed<SpanIdentity>, RawSpan> iterator =
+          spanWindowStore.fetch(
```

Comment:
How is the state store maintaining the order of keys? Do we have to give any ordering function while its creation?
I know, it is reducing our search window, just curious, can we miss any key in between if the order is not the same as the one used above `BYTES_LEXICO_COMPARATOR`?

Reply:
Looks to me that whatever the spanKey added as first and last into the list of `traceStateStore`, that should be the first and last key in the range, isn't it?

Reply:
> How is the state store maintaining the order of keys? Do we have to give any ordering function while its creation?

The keys are stored in sorted order in the store using `BYTES_LEXICO_COMPARATOR`


> can we miss any key in between if the order is not the same as the one used above BYTES_LEXICO_COMPARATOR

yes it is possible

> added as first and last into the list of traceStateStore, that should be the first and last key in the range, isn't it

no, the store keeps the keys in sorted order in the segments which are partitioned by timestamp